File: Cpp/Legacy89DiskKit.Cpp/src/hu_basic_directory_rules.cpp
#include "legacy89diskkit/cpp/hu_basic_directory_rules.hpp"

#include "legacy89diskkit/cpp/hu_basic_name_rules.hpp"

namespace legacy89diskkit::cpp
{
HuBasicFileMetadata HuBasicDirectoryRules::CreateMetadataForWrite(
    const std::vector<std::uint8_t>& data,
    const HuBasicFileAttributes& attributes,
    int start_cluster,
    std::uint16_t load_address,
    std::uint16_t execution_address)
{
    const auto file_type = attributes.is_ascii || (attributes.raw_attributes & 0x0c) != 0
        ? HuBasicFileType::Ascii
        : HuBasicFileType::Binary;

    return HuBasicFileMetadata
    {
        file_type,
        static_cast<std::uint16_t>(data.size()),
        load_address,
        execution_address,
        start_cluster,
        attributes.raw_attributes
    };
}

HuBasicFileEntry HuBasicDirectoryRules::CreateFileEntryForWrite(
    const std::string& file_name,
    const std::vector<std::uint8_t>& data,
    const HuBasicFileAttributes& attributes,
    int start_cluster,
    std::uint16_t load_address,
    std::uint16_t execution_address)
{
    const auto parsed_name = HuBasicNameRules::ParseFileName(file_name);
    const auto end_address = static_cast<std::uint16_t>(load_address + data.size() - 1);

    return HuBasicFileEntry
    {
        parsed_name.file_name,
        parsed_name.extension,
        static_cast<std::uint32_t>(data.size()),
        attributes,
        start_cluster,
        load_address,
        end_address,
        execution_address
    };
}
}

File: CSharp/Legacy89DiskKit.FileSystem.Infrastructure/FileSystem/HuBasic/HuBasicDirectoryRules.cs
using Legacy89DiskKit.FileSystem.Domain.Model;
using Legacy89DiskKit.FileSystem.Infrastructure.HuBasic.Models;

namespace Legacy89DiskKit.FileSystem.Infrastructure.HuBasic;

public static class HuBasicDirectoryRules
{
    public static FileEntry CreateFileEntryForWrite(
        string fileName,
        byte[] data,
        ExtendedFileAttributes attributes,
        int startCluster,
        ushort? loadAddress,
        ushort? executionAddress)
    {
        if (fileName == null) throw new ArgumentNullException(nameof(fileName));
        if (data == null) throw new ArgumentNullException(nameof(data));
        if (attributes == null) throw new ArgumentNullException(nameof(attributes));

        var (name, ext) = HuBasicNameRules.ParseFileName(fileName);
        var fileType = attributes.IsAscii || (attributes.RawAttributes & 0x0C) != 0
            ? HuBasicFileType.Ascii
            : HuBasicFileType.Binary;

        var metadata = new HuBasicFileMetadata(
            fileType,
            false,
            false,
            false,
            false,
            false,
            (ushort)data.Length,
            loadAddress,
            executionAddress,
            startCluster,
            attributes.RawAttributes
        );

        ushort? endAddress = loadAddress.HasValue ? (ushort?)(loadAddress.Value + data.Length - 1) : null;

        return new FileEntry(
            name,
            ext,
            data.Length,
            null,
            DateTime.Now,
            attributes,
            startCluster,
            loadAddress,
            endAddress,
            executionAddress,
            null,
            null,
            metadata
        );
    }
}
